### PR TITLE
Additions to BlockEncode and TensorFlow support

### PR DESCRIPTION
**Pull request by DSGuala**
**Context:**
Adding QSVT to PennyLane

**Description of the Change:**
Finalizes BlockEncode to support TensorFlow and Jax and extends testing.

**Benefits:**
BlockEncode will be more stable and compatible with all frameworks supported by PennyLane.

### CODE REVIEW

**Comment by Jaybsoni** [comment]
The current version does not have support for Tensorflow and only partial support for jax. This will be added to the operation afterwards.

File: tests/math/test_multi_dispatch.py [hunk 1]
```diff
@@ -265,3 +265,16 @@ def test_tf(self):
...
+# class TestNorm:
+#     import tensorflow as tf
+#     import torch
+#     from jax import numpy as jnp
+#
+#     params = [
+#         torch.tensor([5, 6]),
+#         tf.Variable([-1.0, -2.0]),
+#         jnp.array([6, 7]),
+#     ]
```
Comment (soranjh) [inline tests/math/test_multi_dispatch.py:280]: Why this is here?

File: tests/ops/qubit/test_matrix_ops.py [hunk 2]
```diff
@@ -669,9 +685,53 @@ def test_unitary(self, input_matrix, wires):
...
-    # TODO: add tensorflow support
-    # def test_blockencode_tf(self):
-    #     """Test that the BlockEncode operator matrix is correct for tf."""
+    @pytest.mark.parametrize(
+        ("input_matrix", "wires", "msg"),
+        [
+            (
+                [[0, 1], [1, 0]],
+                1,
+                f"Block encoding a \({2} x {2}\) matrix requires a hilbert space of size"
```
Comment (soranjh) [inline tests/ops/qubit/test_matrix_ops.py:694]:
```suggestion
                f"Block encoding a \({2} x {2}\) matrix requires a Hilbert space of size"
```

File: tests/ops/qubit/test_matrix_ops.py [hunk 3]
```diff
@@ -669,9 +685,53 @@ def test_unitary(self, input_matrix, wires):
...
+                    [0.94561648, -0.07621992, -0.1, -0.3],
+                    [-0.07621992, 0.89117368, -0.2, -0.4],
+                ],
+            ),
+            (
+                0.1,
+                range(2),
+                [[0.1, 0.99498744, 0, 0], [0.99498744, -0.1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]],
+            ),
+        ],
```
Comment (soranjh) [inline tests/ops/qubit/test_matrix_ops.py:724]: Should this be defined for each interface separately? I suppose the tests for different interfaces use the same data?

File: pennylane/math/multi_dispatch.py [hunk 4]
```diff
@@ -815,6 +815,18 @@ def expm(tensor, like=None):
     return scipy_expm(tensor)
 
 
+@multi_dispatch()
+def norm(tensor, like=None, **kwargs):
```
Comment (obliviateandsurrender) [inline pennylane/math/multi_dispatch.py:819]: It possibly requires a docstring.

**Review by soranjh — COMMENTED** [review]
Looks good to me, left few questions and comments.

**Review by soranjh — APPROVED** [review]

File: pennylane/ops/qubit/matrix_ops.py [hunk 5]
```diff
@@ -457,33 +473,46 @@ def compute_matrix(*params, **hyperparams):
...
-            col1 = qml.math.vstack([u, qml.math.zeros((r, n + m), like=A)])
-            col2 = qml.math.vstack([qml.math.zeros((n + m, r), like=A), qml.math.eye(r, like=A)])
-            u = qml.math.hstack([col1, col2])
+            col1 = _stack([u, zeros((r, n + m), like=A)], like=interface)
+            col2 = _stack([zeros((n + m, r), like=A), eye(r, like=A)], like=interface)
+            u = _stack([col1, col2], h=True, like=interface)
+
         return u
+
+    def adjoint(self):
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/matrix_ops.py:513]: Does this require unit-tests?

File: pennylane/ops/qubit/matrix_ops.py [hunk 6]
```diff
@@ -457,33 +473,46 @@ def compute_matrix(*params, **hyperparams):
...
+        A = self.parameters[0]
+        return BlockEncode(qml.math.transpose(qml.math.conj(A)), wires=self.wires)
+
+    def label(self, decimals=None, base_label=None, cache=None):
```
Comment (obliviateandsurrender) [inline pennylane/ops/qubit/matrix_ops.py:517]: Does this require unit-tests?

File: tests/math/test_multi_dispatch.py [hunk 7]
```diff
@@ -265,3 +265,16 @@ def test_tf(self):
             out = fn.detach(x)
         jac = t.jacobian(out, x)
         assert jac is None
+
+
+# @pytest.mark.all_interfaces
```
Comment (obliviateandsurrender) [inline tests/math/test_multi_dispatch.py:270]: Please remove commented code.

File: tests/ops/qubit/test_matrix_ops.py [hunk 8]
```diff
@@ -669,9 +685,53 @@ def test_unitary(self, input_matrix, wires):
...
+        [
+            (
+                [[0, 1], [1, 0]],
+                1,
+                f"Block encoding a \({2} x {2}\) matrix requires a hilbert space of size"
+                f" at least \({4} x {4}\). Cannot be embedded in a {1} qubit system.",
+            ),
+        ],
+    )
+    def test_correct_error_message(self, input_matrix, wires, msg):
```
Comment (obliviateandsurrender) [inline tests/ops/qubit/test_matrix_ops.py:699]: Docstring is missing for the test.

File: tests/ops/qubit/test_matrix_ops.py [hunk 9]
```diff
@@ -737,83 +797,103 @@ def test_blockencode_jax(self, input_matrix, wires, output_matrix):
...
+            (0.3, 0),
+            (0.1, range(2)),
+            (
+                [[0.1, 0.2], [0.3, 0.4]],
+                range(2),
+            ),
+            ([[0.1, 0.2, 0.3], [0.3, 0.4, 0.2], [0.1, 0.2, 0.3]], range(3)),
+        ],
+    )
+    def test_adjoint(self, input_matrix, wires):
```
Comment (obliviateandsurrender) [inline tests/ops/qubit/test_matrix_ops.py:892]: Docstring is missing for the test.

File: tests/ops/qubit/test_matrix_ops.py [hunk 10]
```diff
@@ -737,83 +797,103 @@ def test_blockencode_jax(self, input_matrix, wires, output_matrix):
...
+
+    @pytest.mark.tf
+    @pytest.mark.parametrize(
+        ("wires", "input_matrix", "expected_result"),  # expected_results calculated manually
+        [
+            (range(1), pnp.array(0.3), 4 * 0.3),
+            (range(2), pnp.diag([0.2, 0.3]), 4 * pnp.diag([0.2, 0])),
+        ],
+    )
+    def test_blockencode_grad_tf(self, wires, input_matrix, expected_result):
```
Comment (obliviateandsurrender) [inline tests/ops/qubit/test_matrix_ops.py:861]: Docstring is missing for the test.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Leaving a few suggestions

File: tests/math/test_multi_dispatch.py (hunk 1 above)
Comment (Jaybsoni) [reply tests/math/test_multi_dispatch.py]: Missing test was since added

File: tests/math/test_multi_dispatch.py (hunk 7 above)
Comment (Jaybsoni) [reply tests/math/test_multi_dispatch.py]: Missing test was added

File: pennylane/ops/qubit/matrix_ops.py (hunk 6 above)
Comment (Jaybsoni) [reply pennylane/ops/qubit/matrix_ops.py:517]: We should be fine for now 👍🏼

**Pull request merged**